test: run mempool_reorg.py even with wallet disabled

DariusParvin opened the pull request:
Run mempool_reorg.py test even when the wallet is disabled, as discussed in #20078.

As part of this PR I created a new method in `MiniWallet`, `create_self_transfer`, to return a raw tx (without broadcasting it) and its associated utxo.

glozow commented on test/functional/test_framework/wallet.py:
```diff
@@ -67,13 +67,26 @@ def send_self_transfer(self, *, fee_rate=Decimal("0.003"), from_node, utxo_to_sp
+
+        tx_hex, new_utxo = self.create_self_transfer(fee_rate=fee_rate, from_node=from_node, utxo_to_spend=utxo_to_spend, locktime=locktime)
+
+        tx_info = from_node.testmempoolaccept([tx_hex])[0]
```
this is not needed anymore i believe

stackman27 replied:
I believe `tx_info` is being reused used in line 92

stackman27 replied:
Also is there a need to recheck `testmempoolaccept`? Every time we call `send_self_transfer` it checks the mempool acceptance through `create_self_transfer` so, can we just return proper values and remove the additional check?

DariusParvin replied:
good catch! I replaced it with `testmempoolaccept` with `FromHex` to get the txid and wtxid.

glozow commented on test/functional/mempool_reorg.py:
```diff
@@ -46,33 +42,34 @@ def run_test(self):
-        )
-        timelock_tx = self.nodes[0].signrawtransactionwithwallet(timelock_tx)["hex"]
+        utxo = wallet.get_utxo(txid=coinbase_txids[0])
+        timelock_tx, _ = wallet.create_self_transfer(
+            from_node=self.nodes[0],
+            utxo_to_spend=utxo,
+            locktime=self.nodes[0].getblockcount() + 2
+            )
```
nit
```suggestion
        timelock_tx, _ = wallet.create_self_transfer(
            from_node=self.nodes[0],
            utxo_to_spend=utxo,
            locktime=self.nodes[0].getblockcount() + 2
         )
```

glozow commented on test/functional/test_framework/wallet.py:
```diff
@@ -55,8 +55,8 @@ def get_utxo(self, *, txid=''):
+        """Create a tx with the specified fee_rate. Fee may be exact or at most one satoshi higher than needed."""
         self._utxos = sorted(self._utxos, key=lambda k: k['value'])
         utxo_to_spend = utxo_to_spend or self._utxos.pop()  # Pick the largest utxo (if none provided) and hope it covers the fee
         vsize = Decimal(96)
```
Not sure if this is relevant to this PR or previous one, but I'm surprised the vsize is hard-coded? It should be programmatically obtained or asserted to be correct after it's constructed. There's a `get_vsize()` for `CTransaction`s in test_framework.messages

MarcoFalke replied:
There is an assertion `assert_equal(tx_info['vsize'], vsize)`

glozow reviewed (commented):
Concept ACK, very nice!

felipsoarez reviewed (commented):
Concept ACK

MarcoFalke commented on test/functional/mempool_reorg.py:
```diff
@@ -8,36 +8,32 @@
 class MempoolCoinbaseTest(BitcoinTestFramework):
     def set_test_params(self):
         self.num_nodes = 2
+        self.setup_clean_chain = True
```
Generating blocks takes a few seconds in valgrind, so I am thinking if this test may benefit from using `miniwallet.scan_blocks` (to be introduced in #21200)

DariusParvin replied:
I had a go but I don't think it helps in this case since the pre-mined chain doesn't send the coinbase transaction to `self._address`. I made it little bit faster though by only mining 100 blocks instead of 200.

MarcoFalke replied:
It should. Specifically the last 25 mature blocks (including block 100). And the last 25 immature blocks (including block 200), but you don't need those

MarcoFalke replied:
Sorry, there was a bug. If you rebase, the bug should fix itself.

DariusParvin replied:
yes, it worked! I updated the PR to use `miniwallet.scan_blocks` starting from block 76.

stackman27 commented on test/functional/test_framework/wallet.py:
```diff
@@ -67,13 +67,26 @@ def send_self_transfer(self, *, fee_rate=Decimal("0.003"), from_node, utxo_to_sp
+        tx_hex, new_utxo = self.create_self_transfer(fee_rate=fee_rate, from_node=from_node, utxo_to_spend=utxo_to_spend, locktime=locktime)
+
+        tx_info = from_node.testmempoolaccept([tx_hex])[0]
+        self._utxos.append(new_utxo)
```
nit: shouldn't the append be after `sendrawtx` as per [this](https://github.com/bitcoin/bitcoin/pull/21014#discussion_r575909995) comment 
🤔

DariusParvin replied:
switched it around!

stackman27 commented on test/functional/test_framework/wallet.py:
```diff
@@ -67,13 +67,26 @@ def send_self_transfer(self, *, fee_rate=Decimal("0.003"), from_node, utxo_to_sp
+            assert_equal(tx_info['fees']['base'], fee)
+
+        new_utxo = {'txid': tx_info['txid'], 'vout': 0, 'value': send_value}
+        return tx_hex,  new_utxo
```
I'm not very sure the intuition behind this but, can we just add everything in one variable and use key/value pairs to retrieve things? Seems cleaner to me and wouldn't have to include dead `_` every time I create a tx. For instance instead of doing `spend_101_raw, _ = wallet.create_self_transfer(from_node=self.nodes[0], utxo_to_spend=utxo_101)` we can simplt do       `spend_101_raw= wallet.create_self_transfer(from_node=self.nodes[0], utxo_to_spend=utxo_101)['tx_hex']`

DariusParvin replied:
done!

stackman27 reviewed (commented):
Concept ACK. `mempool_reorg` looks really good except, is there a reason why we aren't using `self.log.info(...)`? I always thought of them to be really helpful while running tests. Therefore, i think it'll be really nice to add the logs :)

fanquake commented:
@DariusParvin are you planning on following up with the review comments here?

DariusParvin commented:
@fanquake Yes, thanks for prompting me, I'll push updates this weekend!

DariusParvin commented:
I've addressed the comments, but since there are conflicting changes with #20874, it would probably make sense for that one to get merged first.

DariusParvin commented:
@stackman27 thanks for your review! I added logs instead of comments, as you suggested.

MarcoFalke commented on test/functional/mempool_reorg.py:
```diff
@@ -27,82 +26,89 @@ def skip_test_if_missing_module(self):
         # Use invalidatblock to make all of the above coinbase spends invalid (immature coinbase),
         # and make sure the mempool code behaves correctly.
-        b = [self.nodes[0].getblockhash(n) for n in range(101, 105)]
+        b = [self.nodes[0].getblockhash(n) for n in range(76, 80)]
```
```suggestion
        b = [self.nodes[0].getblockhash(n) for n in range(first_block, first_block+4)]
```

MarcoFalke replied:
Any reason you didn't fix this as well on the latest force push?

DariusParvin replied:
sorry i didn't see this comments, i'll fix it now!

DariusParvin replied:
just pushed it. Thanks for the review

maflcko reviewed (approved):
cr ACK a856630c27cd28186d4e39dd1129f0cae8149261

michaelfolkson commented:
> I've addressed the comments, but since there are conflicting changes with #20874, it would probably make sense for that one to get merged first.

Is this still waiting on #20874? Given Marco has reviewed this and not #20874 presumably this is closer to merging? 😅

michaelfolkson commented:
The test is still skipped when wallet is disabled. I think you need to delete:

` ``
def skip_test_if_missing_module(self):
        self.skip_if_no_wallet()
```

DariusParvin commented:
@michaelfolkson oops! I removed those lines. And yes, since #21762 where `create_self_transfer` was added, this PR is relatively independent of everything else, so it's not waiting on anything else. Cheers

michaelfolkson reviewed (commented):
ACK 65dd482ba8c87443e4c27dc28f2027625ca43005

Reviewed code, test passes (and not skipped) with wallet disabled.

DariusParvin commented on test/functional/mempool_reorg.py:
```diff
@@ -23,86 +22,90 @@ def set_test_params(self):
+        # 1. Direct coinbase spend  :  spend_1
+        # 2. Indirect (coinbase spend in chain, child in mempool) : spend_2 and spend_2_1
+        # 3. Indirect (coinbase and child both in chain) : spend_3 and spend_3_1
         # Use invalidatblock to make all of the above coinbase spends invalid (immature coinbase),
```
thanks! by the way, should I be squashing commits when I making small fixes to the PR or do they get squashed when getting merged?

michaelfolkson replied:
On this project we [squash commits](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits) prior to review and merging. Obviously if there is a reason to have multiple commits then don't squash but in this case I don't think there should be a separate commit for a typo.

michaelfolkson commented:
Re-ACK 163e8881e88a798462a60b5232a6b4ff62e1969a

MarcoFalke commented:
`s/disable/disabled/` in commit message

DariusParvin committed:
test: run mempool_reorg.py even with wallet disabled

- run mempool_reorg.py even when the wallet is not compiled
- add `locktime` argument to `create_self_transfer` and `send_self_transfer`
- use more logs instead of comments

DariusParvin commented:
thanks @MarcoFalke! typo fixed and rebased

MarcoFalke commented:
cr ACK a3f0cbf82ddae2dd83001a9cc3a7948dcfb6fa47
